aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -0,0 +1,516 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/core/Tensor.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/Parallel.h>
+#include <ATen/cpu/vec/vec.h>
+#include <ATen/cpu/vec/functional.h>
+#include <ATen/native/CPUBlas.h>
+#include <ATen/native/cpu/utils.h>
+#include <ATen/native/transformers/attention.h>
+#include <ATen/native/transformers/sdp_utils_cpp.h>
+#include <c10/util/irange.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#else
+#include <ATen/ops/empty.h>
+#endif
+
+#if AT_MKL_ENABLED()
+#include <mkl.h>
+#endif
+
+namespace at::native {
+
+namespace {
+
+using fVec = vec::Vectorized<float>;
+static auto f_vec_size = fVec::size();
+
+inline void _store(
```

Comment:
we can use `map` from functional to write most of the logic, it would be more clean. as long as the buffer size is L1 hit, the perf doesn't make much difference.